### PR TITLE
feat: refactor client state machine

### LINKED ISSUE #1181: Convert ad-hoc state machine to futures state machine lib

### PR DESCRIPTION
**Pull request by bbangert**
Switch to using the futures state machine to separate out client
state from the implementation.

Closes #1181

### CODE REVIEW

File: autopush_rs/src/client.rs [hunk 1]
```diff
@@ -112,747 +211,907 @@ impl ClientFlags {
...
-            if let ClientState::Done = self.state {
-                return Ok(().into());
+    #[state_machine_future(transitions(IncrementStorage, CheckStorage, WaitingForDropUser,
+                                       WaitingForMigrateUser, WaitForInput))]
+    DetermineAck { data: AuthClientData<T> },
+
+    #[state_machine_future(transitions(DetermineAck, SendThenWait, WaitForInput,
+                                       WaitingForRegister, WaitingForUnregister,
+                                       WaitingForDelete))]
+    WaitForInput { data: AuthClientData<T> },
```
Comment (pjenvey) [inline autopush_rs/src/client.rs]: WaitForInput -> WaitingForInput

Comment (bbangert) [reply autopush_rs/src/client.rs]: :grinning:

**Comment by alexcrichton** [comment]
I may not personally be of much use at the technical portions of this but it seems pretty nice! I haven't used this state machine library before but it looks quite nifty

File: autopush_rs/src/client.rs [hunk 2]
```diff
@@ -112,747 +211,837 @@ impl ClientFlags {
...
-    pub fn shutdown(&mut self) {
-        // If we made it past hello, do more cleanup
-        let mut webpush = match self.webpush.take() {
-            Some(webpush) => webpush,
-            None => return
+    fn poll_await_session_complete<'a>(
+        session_complete: &'a mut RentToOwn<'a, AwaitSessionComplete<T>>,
+    ) -> Poll<AfterAwaitSessionComplete, Error> {
+        // xxx: handle error cases with maybe a log message?
+        let _error = {
```
Comment (jrconlin) [inline autopush_rs/src/client.rs]: Add a filtering log message here?

File: autopush_rs/src/client.rs [hunk 3]
```diff
@@ -112,747 +211,837 @@ impl ClientFlags {
...
+
+    #[state_machine_future(ready)]
+    UnAuthDone(()),
+
+    #[state_machine_future(error)]
+    UnAuthClientStateError(Error),
 }
 
-impl<T> ClientData<T>
+impl<T> PollUnAuthClientState<T> for UnAuthClientState<T>
```
Comment (jrconlin) [inline autopush_rs/src/client.rs]: Add a comment noting that Impl and method names have a required formula? This might be easy to screw up in the future?

**Pull request merged**